crates/storage-query-datafusion/src/table_util.rs

```diff
@@ -68,3 +69,48 @@ pub(crate) trait Builder {
 
     fn finish(self) -> datafusion::common::Result<RecordBatch>;
 }
+
+pub(crate) struct BatchSender<B: Builder> {
+    projection: SchemaRef,
+    builder: B,
+    tx: Sender<datafusion::error::Result<RecordBatch>>,
+}
+
+impl<B: Builder> BatchSender<B> {
+    pub(crate) fn new(
+        projection: SchemaRef,
+        tx: Sender<datafusion::error::Result<RecordBatch>>,
+    ) -> Self {
+        let builder = B::new(projection.clone());
+        Self {
+            projection,
+            builder,
+            tx,
+        }
+    }
+
+    pub(crate) fn send(
+        &mut self,
+    ) -> Result<(), tokio::sync::mpsc::error::SendError<datafusion::error::Result<RecordBatch>>>
+    {
+        let old_builder = std::mem::replace(&mut self.builder, B::new(self.projection.clone()));
+        let result = old_builder.finish();
+        self.tx.blocking_send(result)
+    }
+
+    pub(crate) fn builder_mut(&mut self) -> &mut B {
+        &mut self.builder
+    }
+
+    pub(crate) fn num_rows(&self) -> usize {
+        self.builder.num_rows()
+    }
+}
+
+impl<B: Builder> Drop for BatchSender<B> {
```

Comment:
this is really critical to this approach, as it means the for each doesn't have to be called with None or something on the last iteration. instead, we just send what's left on Drop